internal/vcs/metadata.go

```diff
@@ -0,0 +1,244 @@
+package vcs
+
+import (
+	"encoding/base64"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/object"
+	"github.com/go-git/go-git/v5/plumbing/transport/http"
+	"github.com/tidwall/gjson"
+)
+
+var (
+	// envToProviderMap specifies unique env vars that determine the CI system being used and maps
+	// them to the corresponding function to find vcs metadata.
+	envToProviderMap = map[string]GetMetadataFunc{
+		"GITHUB_ACTIONS": getGithubMetadata,
+		"GITLAB_CI":      getGitlabMetadata,
+	}
+
+	StubMetadata = Metadata{
+		Branch: Branch{
+			Name: "stub-branch",
+		},
+		Commit: Commit{
+			SHA:       "stub-sha",
+			Author:    "stub-author",
+			Timestamp: 12345,
+			Message:   "stub-message",
+		},
+	}
+)
+
+// GetMetadataFunc defines a function that fetches data for a given vcs implementation.
+type GetMetadataFunc func() (Metadata, error)
+
+// GetMetadata fetches vcs metadata for the given environment. GetMetadata will attempt to try and find
+// a GetMetadataFunc for the environment using env variables to determine the CI system. If GetMetadata
+// cannot determine the CI system it falls back to getting the metadata from the local git filesystem.
+func GetMetadata() (Metadata, error) {
+	if isTest() {
+		return StubMetadata, nil
+	}
+
+	for e, f := range envToProviderMap {
+		_, ok := os.LookupEnv(e)
+		if ok {
+			return f()
+		}
+	}
+
+	return getLocalGitMetadata()
+}
+
+func isTest() bool {
+	return os.Getenv("INFRACOST_ENV") == "test" || strings.HasSuffix(os.Args[0], ".test")
+}
+
+func getGitlabMetadata() (Metadata, error) {
+	m, err := getLocalGitMetadata()
+	if err != nil {
+		return m, fmt.Errorf("github metdata error, could not fetch initial metadata from local git %w", err)
+	}
+
+	if m.Branch.Name == "HEAD" {
+		m.Branch.Name = os.Getenv("CI_MERGE_REQUEST_SOURCE_BRANCH_NAME")
+	}
+
+	m.Pipeline = &Pipeline{ID: os.Getenv("CI_JOB_ID")}
+	m.PullRequest = &PullRequest{
+		VCSProvider:  "gitlab-ci",
```

Comment:
Shouldn't the VCS provider be called just `gitlab`?
```suggestion
		VCSProvider:  "gitlab",
```

Reply:
👍